### ISSUE TITLE
Use of angle brackets around file names for include statements

### ISSUE THREAD

**Issue by elfring** [VcDevel/Vc#74]
Would you like to [replace any double quotes](https://github.com/VcDevel/Vc/blob/ed3869b789f4dc5dc3e995b2fde4ceef2054327a/include/Vc/Vc#L22 "Check header file inclusion style") by [angle brackets around file names for include statements](http://stackoverflow.com/questions/21593/what-is-the-difference-between-include-filename-and-include-filename "What is the difference between #include <filename> and #include “filename”?")?

**Comment by mattkretz** [comment]
No, why? I want relative includes here. Is there any crazy interpretation of `#include ""` that I need to be aware of? At this point Vc only supports GCC, clang, and ICC so I'm pretty sure double quotes do what I expect them to do.

**Comment by elfring** [comment]
I suggest to reconsider the consequences of [the following wording](http://c0x.coding-guidelines.com/6.10.2.html "Source file inclusion") from [the section "16.2 Source file inclusion"](http://open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3797.pdf "C++ working draft") in the standard specification for the programming language "C++".
```
…
The named source file is searched for in an implementation-defined manner. If this search is not supported, or if the search fails, the directive is reprocessed as if it read

#include <h-char-sequence> new-line
…
```

* How do you think about to avoid "a duplicated file search" here?
* Would you like to [restrict the searched directories](https://gcc.gnu.org/onlinedocs/cpp/Include-Syntax.html "Include syntax description") for header files of your software?

**Comment by mattkretz** [comment]
* what "duplicated file search" do you mean? The `#include ""` statements are all for relative includes, so they never fall back to `-I` paths (when installed - in the Vc sources they actually do fall back to `-I`, but then again the include paths are under my control, so that's fine)
* "Would you like to restrict the searched directories for header files of your software?"
Not sure what you're suggesting, but yes, that's what I achieving here. I.e. the double quote includes never pick up files specified by `-I` paths.

**Comment by elfring** [comment]
>… the double quote includes never pick up files specified by -I paths.

I find your view interesting. - I would prefer other build system settings for reusing your software.

**Comment by mattkretz** [comment]
> I would prefer other build system settings for reusing your software

I don't understand. What build system settings do you mean?

**Comment by js75** [comment]
Why would that be safer in Vc's case? Unless you are planning to reorganize the source tree I don't really see your problem with the way header files are currently included.

**Comment by elfring** [comment]
There are different opinions about the handling of the involved implementation-defined behaviour.
* Will header files be also searched outside the specified include directories if double quotes are used for the discussed preprocessor statement?
* Is there a speed difference measurable if a file is not found there and the search will be retried with "the angle brackets inclusion method"?

**Comment by mattkretz** [comment]
@elfring I still can't follow what you're trying to tell me.

Again here's what's happening, maybe you can write a concrete statement what is wrong about it:
* An `#include "file.h"` is resolved by all supported compilers as looking first for `file.h` in the same directory as the file containing the `#include` directive.
* If the Vc headers are installed then the compiler will find `file.h` in the relative path in all cases (everything else would actually be a bug). I.e. a single `stat` suffices.
* If the Vc headers are not installed then a few of the relative includes are not found and the compiler falls back to `-I` paths (there are only two `-I` switches on the command line). There might be a compilation performance issue here, but I don't care much, because what counts is when Vc is installed.

**Comment by elfring** [comment]
>An `#include "file.h"` is resolved by all supported compilers as looking first for `file.h` in the same directory as the file containing the `#include` directive.

Would you like to be able to distinguish the installed class library and a version which was "accidentally" copied into the source directory of an other software project?


>There might be a compilation performance issue here, …

Do any other developers or software packagers care eventually a bit more about this detail?

**Comment by mattkretz** [comment]
I would be so happy if you could stop asking vague questions and instead say what issue you're seeing that needs to be fixed. Otherwise I fear we're wasting our time here.

> Would you like to be able to distinguish the installed class library and a version which was "accidentally" copied into the source directory of an other software project?

I don't understand what you're trying to suggest.

> Do any other developers or software packagers care eventually a bit more about this detail?

If a packager cares more about a 1% packaging speed increase than correct relative includes in the dev packages he is creating then I don't care about his opinion. A library's focus is on its users not on its packagers.

**Comment by mattkretz** [comment]
@elfring BTW, I would hope you are aware of [KDE's policy on this](https://techbase.kde.org/Policies/Library_Code_Policy#Getting_.23includes_right). Could you please make a case for your quest of submitting this issue to arbitrary projects by addressing the rationale in KDE's policy? (I wrote that text, so that's my background on this issue...)

**Comment by elfring** [comment]
* Do you care then for example why [the boost class library](http://www.boost.org/ "Boost C++ libraries") is using angle brackets in include statements for corresponding header files?
* When do you categorise a development situation in this way that the headers of your software belong also to [just another external library](https://techbase.kde.org/Policies/Library_Code_Policy#As_application_developer "Selection of inclusion style")?

**Comment by paulelsner** [comment]
In my opinion using angle brackets will not solve the problem, it could
make it even worse.
I think Markus is talking about name conflicts when you have multiple
libraries in you project and they could have header files with the same
name.

In the past I'v already had a conflict there. In my library I shipped a
public header named config.h. In the public header the user had to include
I used #include "config.h".
Everythink fine to far. Than a user was also using WebRTC which also ships
a config.h. In WebRTC some files use #include <config.h>. This leads to the
error that my header becomes included, because it was the first one in the
compilers search path, than the WebRTCs own one.

The workaround for this issue was to move my config.h header in a subfolder
"detail" and include this way: #include "detail/config.h". Better solution
would be to fix WebRTC to not use angle brackets.

In short: using angle brackets in public headers to include the file of the
own library is wrong.

Hope that helps,

Paul

2015-11-19 18:45 GMT+01:00 Markus Elfring <notifications@github.com>:

**Comment by elfring** [comment]
>I think Markus is talking about name conflicts when you have multiple libraries in you project and they could have header files with the same name.

Not really. - But I was concerned about similar consequences.


>In the past I'v already had a conflict there.

Thanks for your concrete example.


>In short: using angle brackets in public headers to include the file of the own library is wrong.

I disagree to your conclusion for this implementation detail. I suggest to consider the possibility that a software build system can eventually determine more appropriate settings there.
Should any more libraries refer to their header files with a subdirectory in their include path?

**Comment by mattkretz** [comment]
@elfring Again more questions. This is not helping.

> Do you care then for example why the boost class library is using angle brackets in include statements for corresponding header files?

Without a convincing rationale I don't care, why should I?

> When do you categorise a development situation in this way that the headers of your software belong also to just another external library?

Either I don't understand the question or it doesn't make sense. Here's what I read: “what about headers that are part of my software and belong to another library?” How does that scenario even make sense? Please give a concrete example of the issue that needs to be fixed.

@paulelsner wrote:
> I think Markus is talking about name conflicts when you have multiple libraries in you project and they could have header files with the same name.

Actually that's exactly why you should not use angle brackets exclusively (as the [KDE Policy document explains](https://techbase.kde.org/Policies/Library_Code_Policy#Getting_.23includes_right)). Therefore @elfring 's quest for the removal of quoted includes is leading to unnecessary breakage and more brittle `-I` orders for library users.

**Comment by mattkretz** [comment]
@elfring I challenge you to try out your suggestion on e.g. `include/Vc/vector.h` and see whether you can make it work.
